### PR TITLE
Add telemetry for SDK usage from DBR

### PR DESCRIPTION
**Pull request by shreyas-goenka**
## Changes
This PR adds telemetry through the user agent for DBR usage from the SDK. We use the same mechanism and key as the Python SDK to add this to the user agent (PR: https://github.com/databricks/databricks-sdk-py/pull/287).

## Tests
Using unit tests

### CODE REVIEW

File: useragent/user_agent.go [hunk 1]
```diff
@@ -85,13 +85,27 @@ func (u info) String() string {
...
+	}
+	if !isAlphanum(value) && !isSemVer(value) {
+		return fmt.Errorf("expected user agent value for key %q to be alphanumeric or semver: %s", key, value)
+	}
+	return nil
+}
+
 // With always uses the latest value for a given alphanumeric key.
 // Panics if key or value don't satisfy alphanumeric or semver format.
 func (d data) With(key, value string) data {
```
Comment (nfx) [inline useragent/user_agent.go:107]: why not a simple `if` statement in the `With` method?

**Review by nfx — CHANGES_REQUESTED** [review]

File: useragent/user_agent.go (hunk 1 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go:107]: It's cleaner this way IMO since "runtime" is the exception case.  We still want to add the key when it's "runtime", meaning the check ends up being something like:
```
 if key != "runtime" && !isAlphanum(value) && !isSemVer(value) {}
```

Comment (shreyas-goenka) [reply useragent/user_agent.go:107]: Also better error messages and can unit test this way. With the panic stack trace it's difficult to tell where exactly the validation failed.

File: useragent/patterns.go [hunk 2]
```diff
@@ -36,3 +36,11 @@ func matchAlphanumOrSemVer(s string) error {
 	}
 	return fmt.Errorf("invalid alphanumeric or semver string: %s", s)
 }
+
+func isAlphanum(s string) bool {
+	return regexpAlphanum.MatchString(s)
```
[the remark this answers is not in the document]
Comment (shreyas-goenka) [reply useragent/patterns.go:41]: done

File: config/api_client.go [hunk 3]
```diff
@@ -64,6 +66,20 @@ func (c *Config) NewApiClient() (*httpclient.ApiClient, error) {
...
+				// Detect if the SDK is being run in a Databricks Runtime.
+				v, ok := os.LookupEnv("DATABRICKS_RUNTIME_VERSION")
+				if !ok {
+					return nil
+				}
+				// Replace any slashes and spaces with dashes, just in case.
+				v = strings.ReplaceAll(v, "/", "-")
+				v = strings.ReplaceAll(v, " ", "-")
+				// Add the detected Databricks Runtime version to the user agent
+				ctx := useragent.InContext(r.Context(), "runtime", v)
```
[the remark this answers is not in the document]
Comment (shreyas-goenka) [reply config/api_client.go:79]: We have to call it "runtime" in this case because the Python SDK already calls it runtime, and we don't want to diverge from that convention.
see: https://github.com/databricks/databricks-sdk-py/pull/287

The string constant part is done!

File: useragent/user_agent.go [hunk 4]
```diff
@@ -79,19 +86,48 @@ type info struct {
 	Value string
 }
 
+// Sanitize the user agent value. This is useful when the value is not ensured to be
+// to be valid at compile time. Having this sanitization then ensures downstream
+// applications can correctly parse the full user agent header, by making sure
+// characters like '/' and ' ' are not present in the value.
+func Sanitize(s string) string {
+	allowList := func(r rune) rune {
+		if r == '.' || unicode.IsLetter(r) || unicode.IsDigit(r) {
```
Comment (pietern) [inline useragent/user_agent.go:95]: This includes everything in the "letter" class, which includes modifiers, accents, and more that doesn't match the regexp for alphanumerical values.

To guarantee that sanitize produces a valid value it would need to use the same set of characters as the characters in the alphanumeric regexp and map all non-matching characters to the placeholder.

File: useragent/user_agent.go [hunk 5]
```diff
@@ -79,19 +86,48 @@ type info struct {
...
 	return fmt.Sprintf("%s/%s", u.Key, u.Value)
 }
 
 type data []info
 
+// Validate the key value pair being set in the user agent. Error if invalid.
+func validate(key, value string) error {
+	// DBR versions as set in the `DATABRICKS_RUNTIME_VERSION` environment variable
+	// are not valid semver strings. Eg: 15.1 or client.0. Thus we allow arbitrary
+	// values for the runtime key.
```
Comment (pietern) [inline useragent/user_agent.go:114]: We could broaden `regexpAlphanum` class to include `.`. That seems less brittle than the current approach.

Are there other examples of values that we need to allow?

File: useragent/user_agent.go (hunk 1 above)
Comment (pietern) [reply useragent/user_agent.go:129]: A refresh of the error messages sounds good, but special casing `runtime` does not.

File: useragent/user_agent_test.go [hunk 6]
```diff
@@ -60,3 +60,22 @@ func TestFromContext_Custom(t *testing.T) {
...
+	assert.NoError(t, validate("runtime", "whatever#!@"))
+	assert.NoError(t, validate("abc", "123"))
+	assert.NoError(t, validate("abc", "1.1.1"))
+}
+
+func TestUserAgentNormalizeString(t *testing.T) {
+	assert.Equal(t, "abc123", Sanitize("abc123"))
+	assert.Equal(t, "1-2-3-4-5-6-7-8-", Sanitize("1@2#3?4,5/6!7 8 "))
+	assert.Equal(t, "1.2.3", Sanitize("1.2.3"))
+	assert.Equal(t, "client.0", Sanitize("client.0"))
```
Comment (pietern) [inline useragent/user_agent_test.go:80]: Please test that sanitized values match `isAlphanum`, and include Unicode characters, emoji's, etc.

File: useragent/patterns_test.go [hunk 7]
```diff
@@ -8,25 +8,54 @@ import (
...
+	assert.True(t, isAlphanum("foo-bar"))
+
 	assert.Error(t, matchAlphanum("foo bar"))
+	assert.False(t, isAlphanum("foo bar"))
+
 	assert.Error(t, matchAlphanum("foo/bar"))
+	assert.False(t, isAlphanum("foo/bar"))
 }
 
 func TestMatchAlphanumOrSemVer(t *testing.T) {
```
Comment (pietern) [inline useragent/patterns_test.go:46]: I don't think this function is used.

File: useragent/user_agent.go [hunk 8]
```diff
@@ -79,19 +86,48 @@ type info struct {
...
+	if key == RuntimeKey {
+		return nil
+	}
+	if !isAlphanum(key) {
+		return fmt.Errorf("expected user agent key to be alphanumeric: %s", key)
```
Comment (pietern) [inline useragent/user_agent.go:119]: Use `%q` for key so that it is quoted as well (it's not alphanumeric so it could be anything).

File: useragent/user_agent.go [hunk 9]
```diff
@@ -79,19 +86,48 @@ type info struct {
...
+	}
+	if !isAlphanum(value) && !isSemVer(value) {
```
Comment (pietern) [inline useragent/user_agent.go:121]: Did you mean to use `matchAlphanumOrSemVer`?

File: client/client_test.go [hunk 10]
```diff
@@ -320,6 +320,64 @@ GET /a
...
 }
 
+func TestUserAgentForDBR(t *testing.T) {
+	goVersion := strings.TrimPrefix(runtime.Version(), "go")
+	cicdHeader := ""
+	if useragent.CiCdProvider() != "" {
+		cicdHeader = fmt.Sprintf(" cicd/%s", useragent.CiCdProvider())
+	}
+	t.Setenv("DATABRICKS_RUNTIME_VERSION", "15.1")
+	expectedUserAgent := `unknown/0.0.0 databricks-sdk-go/` + version.Version + ` go/` + goVersion + ` os/` + runtime.GOOS + ` auth/pat` + cicdHeader + ` runtime/15.1`
```
Comment (pietern) [inline client/client_test.go:330]: Can you check `assert.Contains(t, " runtime/15.1", userAgent)` instead?

The other details seem irrelevant to the test.

File: client/client_test.go [hunk 11]
```diff
@@ -320,6 +320,64 @@ GET /a
...
+				Body:       io.NopCloser(strings.NewReader(`{}`)),
+				Request:    r,
+			}, nil
+		}),
+	})
+	require.NoError(t, err)
+	err = c.Do(context.Background(), "GET", "/a", nil, nil, nil)
+	assert.Equal(t, userAgent, expectedUserAgent)
+	require.NoError(t, err)
+}
```
Comment (pietern) [inline client/client_test.go:379]: This test is equivalent to the one above, except for the value of the env var.

File: useragent/user_agent.go (hunk 1 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go:130]: Removed the exception for `runtime`

File: useragent/user_agent.go (hunk 4 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go]: Good catch. Did some research and it seems UTF8 characters are not supported in general in HTTP headers. The recommended approach seems to be to stick to ASCII characters in the HTTP headers. 

I have changed the validation to allow alphanumeric, `-`, `_` AND `.` as valid characters.

File: useragent/user_agent.go (hunk 5 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go]: Done. We don't need any other characters at the moment, but we can expand this set as and when needed.

File: useragent/user_agent_test.go (hunk 6 above)
Comment (shreyas-goenka) [reply useragent/user_agent_test.go:113]: Done

File: useragent/user_agent.go (hunk 8 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go]: Done

File: useragent/user_agent.go (hunk 9 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go]: Now we use a single function, `isValid`

File: client/client_test.go (hunk 10 above)
Comment (shreyas-goenka) [reply client/client_test.go]: Done

File: client/client_test.go (hunk 11 above)
Comment (shreyas-goenka) [reply client/client_test.go:378]: Good point. Added test casing.

File: useragent/patterns.go [hunk 12]
```diff
@@ -14,25 +14,28 @@ const (
...
-		return nil
-	}
-	return fmt.Errorf("invalid alphanumeric string: %s", s)
+// Alphanumeric characters, hyphen, underscore, and period. This is the subset of
+// characters that we allow in user agent keys and values. This is to ensure that
+// downstream applications can correctly parse the full user agent header.
+//
+// NOTE: HTTP headers in general only work well with ASCII characters. see:
+// https://stackoverflow.com/questions/4400678/what-character-encoding-should-i-use-for-a-http-header
+var validChars = `0-9A-Za-z_\-\.`
```
Comment (pietern) [inline useragent/patterns.go:30]: Please initialize the regexps you wish to use as globals as well. This string (which can be a const) is not useful outside of regexps. It can also include the brackets because it is always a character class.

The regexp compile adds overhead for every call (multiple calls per request) that we can amortize by initializing it only once.

File: useragent/user_agent.go [hunk 13]
```diff
@@ -99,13 +106,29 @@ func (u info) String() string {
 
 type data []info
 
+// Validate the key value pair being set in the user agent. Error if invalid.
+func validate(key, value string) error {
+	if !isValid(key) {
+		return fmt.Errorf("expected user agent key to be alphanumeric: %q", key)
+	}
+	if !isValid(value) {
```
Comment (pietern) [inline useragent/user_agent.go:114]: This no longer checks for semver strings, which can already be in use.

I believe the semver string is a superset of the current alphanum class, so that can be an issue.

Also, I don't think you need the `isValid` function and can avoid the additional function definition.

File: useragent/user_agent_test.go [hunk 14]
```diff
@@ -74,3 +74,56 @@ func TestFromContext_Custom(t *testing.T) {
...
+	assert.Equal(t, "foo-bar", Sanitize("foo bar"))
+
+	assert.False(t, isValid("foo/bar"))
+	assert.Equal(t, "foo-bar", Sanitize("foo/bar"))
+
+	assert.False(t, isValid("foo:)bar"))
+	assert.Equal(t, "foo--bar", Sanitize("foo:)bar"))
+
+	assert.False(t, isValid("foo😊bar"))
+	assert.Equal(t, "foo-bar", Sanitize("foo😊bar"))
```
Comment (pietern) [inline useragent/user_agent_test.go:128]: This is hard to read; both the valid cases and the invalid but sanitizes cases can be a slice.

File: config/api_client.go [hunk 15]
```diff
@@ -60,7 +61,19 @@ func (c *Config) NewApiClient() (*httpclient.ApiClient, error) {
...
+				*r = *r.WithContext(ctx) // replace request
+				return nil
+			},
+			func(r *http.Request) error {
+				// Detect if the SDK is being run in a Databricks Runtime.
+				v, ok := os.LookupEnv("DATABRICKS_RUNTIME_VERSION")
+				if !ok {
+					return nil
+				}
+				v = useragent.Sanitize(v)
```
Comment (pietern) [inline config/api_client.go:74]: Please also move this to `useragent/runtime.go` so we can cache the output just like we do for CI/CD above.

Comment (shreyas-goenka) [reply config/api_client.go]: done

File: useragent/user_agent_test.go (hunk 14 above)
Comment (shreyas-goenka) [reply useragent/user_agent_test.go]: done

File: useragent/user_agent.go (hunk 13 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go]: done

File: useragent/patterns.go (hunk 12 above)
Comment (shreyas-goenka) [reply useragent/patterns.go]: removed

File: useragent/user_agent.go [hunk 16]
```diff
@@ -99,6 +106,22 @@ func (u info) String() string {
...
+//
+// Note: Semver strings are comprised of alphanumeric characters, hyphens, periods
+// and plus signs. This function will not remove these characters.
+// see:
+// 1. https://semver.org/#spec-item-9
+// 2. https://semver.org/#spec-item-10
+func Sanitize(s string) string {
+	return regexp.MustCompile(`[^0-9A-Za-z_\.\+-]`).ReplaceAllString(s, "-")
+}
+
```
Comment (pietern) [inline useragent/user_agent.go:124]: Please move to `patterns.go` and keep the regexp definition close to the one defining the regexp for alphanum.

More likely they will drift otherwise.

The regexp can also be statically initialized instead of recompiled on every call (per my earlier comment).

File: useragent/runtime.go [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+func Runtime() string {
+	if DisableRuntimeCaching {
+		return getRuntimeVersion()
+	}
+
+	runtimeOnce.Do(func() {
+		runtimeVersion = getRuntimeVersion()
+	})
+	return runtimeVersion
+}
```
Comment (pietern) [inline useragent/runtime.go:29]: Like the CI/CD test, this one can have a companion test where you verify that a few select runtime version strings are correctly returned/sanitized and are valid. Then the e2e test can just verify that it is set and contains something and not iterate over all options.

Also removes the need for exported symbol for `DisableRuntimeCaching`.

**Review by pietern — COMMENTED** [review]
Please use newlines in all table tests. It is hard to read the cases that are covered otherwise.

File: useragent/user_agent.go (hunk 16 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go]: done

File: useragent/runtime.go (hunk 17 above)
Comment (shreyas-goenka) [reply useragent/runtime.go:21]: done

File: useragent/user_agent.go (hunk 16 above)
Comment (shreyas-goenka) [reply useragent/user_agent.go]: done

File: useragent/runtime.go (hunk 17 above)
Comment (shreyas-goenka) [reply useragent/runtime.go:21]: The end to end test is failing because of the caching. Does it make sense to keep the exported flag?

File: useragent/user_agent_test.go [hunk 18]
```diff
@@ -74,3 +74,30 @@ func TestFromContext_Custom(t *testing.T) {
 func TestDefaultsAreValid(t *testing.T) {
 	WithProduct(productName, productVersion)
 }
+
+func TestSanitize(t *testing.T) {
```
Comment (pietern) [inline useragent/user_agent_test.go:78]: Nit: this belongs in `patterns_test.go`.

File: useragent/runtime.go (hunk 17 above)
Comment (pietern) [reply useragent/runtime.go:21]: What do we do for the CI/CD one?

Comment (shreyas-goenka) [reply useragent/runtime.go:21]: We originally did had the e2e test for the CI/CD one. Also the tests are run from GitHub actions so another one of our tests as a side effect asserted on the CICD useragent. (`TestNonJSONResponseIncludedInError`)

Comment (shreyas-goenka) [reply useragent/runtime.go:30]: Done. Added a `ClearCache` method and also added an e2e test for the cicd user agent.

File: useragent/user_agent_test.go (hunk 18 above)
Comment (shreyas-goenka) [reply useragent/user_agent_test.go]: done

File: client/client_test.go [hunk 19]
```diff
@@ -320,6 +321,99 @@ GET /a
...
+		},
+		"tf-cloud": {
+			"TFC_RUN_ID": "",
+		},
+	}
+
+	for ci, envVars := range ciToEnv {
+		t.Run(ci, func(t *testing.T) {
+			env.CleanupEnvironment(t)
+			useragent.ClearCache()
```
Comment (pietern) [inline client/client_test.go:402]: What's the rationale for a `t.Run` here but not in the other test?

File: client/client_test.go [hunk 20]
```diff
@@ -320,6 +321,99 @@ GET /a
...
+	require.NoError(t, err)
+
+	err = c.Do(context.Background(), "GET", "/a", nil, nil, nil)
+	require.NoError(t, err)
+
+	return userAgent
+}
+
+func TestUserAgentForDBR(t *testing.T) {
+	for _, dbrVersion := range []string{"client.0", "client.1", "15.5", "15.5.0", "13.3"} {
```
Comment (pietern) [inline client/client_test.go:350]: Please include newline separators here as well for readability.

Btw, doesn't this duplicate the tests done in the `useragent` package?

File: useragent/runtime_test.go [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+)
+
+func TestUserAgentRuntime(t *testing.T) {
+	for _, tc := range []string{
+		"1.2.3",
+		"0.0.0-dev+2e014739024a",
+		"client.0",
+		"foo",
+		"15.0",
+		"13.3",
```
Comment (pietern) [inline useragent/runtime_test.go:16]: Can you include an emoji/invalid character here as well?

I think this test passes if you remove the `Sanitize` call from the function.

**Review by pietern — APPROVED** [review]

File: client/client_test.go (hunk 20 above)
Comment (shreyas-goenka) [reply client/client_test.go:350]: This is more robust since it does not depend upon the implementation details of the validation logic and tests it end to end. In the other test we call the `match...` function manually to verify validation does not fail.

File: client/client_test.go (hunk 19 above)
Comment (shreyas-goenka) [reply client/client_test.go:402]: We should have `t.Run` at both places. That makes for better error logs. Updated the PR.

File: client/client_test.go (hunk 20 above)
Comment (pietern) [reply client/client_test.go]: That's fine, but could you then also include a garbage value to test that it doesn't panic/fail.

Comment (shreyas-goenka) [reply client/client_test.go]: Added a couple of cases that need sanitization.

File: useragent/runtime_test.go (hunk 21 above)
Comment (shreyas-goenka) [reply useragent/runtime_test.go:16]: done
